Lib/test/test_capi/test_opt.py

```diff
@@ -1953,6 +1953,61 @@ def testfunc(n):
         self.assertNotIn("_GUARD_NOS_INT", uops)
         self.assertNotIn("_GUARD_TOS_INT", uops)
 
+    def test_call_len_known_length_small_int(self):
+        def testfunc(n):
+            x = 0
+            for _ in range(n):
+                a = (1, 2, 3, 4, 5)
+                if len(a) == 5:
+                    x += 1
+            return x
+
+        res, ex = self._run_with_optimizer(testfunc, TIER2_THRESHOLD)
+        self.assertEqual(res, TIER2_THRESHOLD)
+        self.assertIsNotNone(ex)
+        uops = get_opnames(ex)
+        # When the length is < _PY_NSMALLPOSINTS, the len() call is replaced
+        # with just an inline load.
+        self.assertNotIn("_CALL_LEN", uops)
+        self.assertNotIn("_POP_CALL_ONE_LOAD_CONST_INLINE_BORROW", uops)
+        self.assertNotIn("_POP_CALL_LOAD_CONST_INLINE_BORROW", uops)
+        self.assertNotIn("_POP_TOP_LOAD_CONST_INLINE_BORROW", uops)
+
+    def test_call_len_known_length(self):
+        def testfunc(n):
+            x = 0
+            for _ in range(n):
+                a = (1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
```

Comment:
I don't know how else to create a tuple with more than `_PY_NSMALLPOSINTS` elements in a way that the optimizer will know the length :laughing: (`_CALL_TUPLE_1` doesn't have any optimization for instance)

Reply:
Class attributes are cached as constants, so maybe something like this would work?

```py
class C:
    t = tuple(range(300))

x = 0
for _ in range(n):
    if len(C.t) == 300:
        x += 1
```